// @chainlink/belt/lib/src/services/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shelljs_1 = require("shelljs");
const utils_1 = require("./utils");
/**
 * Load a validated configuration file in JSON format for app configuration purposes.
 *
 * @param path The path relative to the current working directory to load the configuration file from.
 */
function load(path) {
    let json;
    try {
        json = utils_1.getJsonFile(path);
    }
    catch (e) {
        throw Error(`Could not load config at "${path}".\n\n${e}`);
    }
    assertAppConfig(json);
    return json;
}
exports.load = load;
function assertAppConfig(json) {
    function assertStr(val, prop) {
        if (typeof val !== 'string') {
            throw Error(`Expected value of config.${prop} to be a string\nGot: ${val}`);
        }
    }
    function assertDir(val, prop) {
        assertStr(val, prop);
        if (!shelljs_1.test('-d', val)) {
            throw Error(`Expected value of config.${prop} to be a directory\nGot: ${val}`);
        }
    }
    function assertBool(val, prop) {
        if (typeof val !== 'boolean') {
            throw Error(`Expected value of config.${prop} to be a boolean\nGot: ${val}`);
        }
    }
    function assertCompilerSettings(val) {
        if (!val || typeof val !== 'object') {
            throw Error(`Expected value of config.compilerSettings to be an object\nGot:${val}`);
        }
        const compilerSettings = val;
        if (!compilerSettings.versions ||
            typeof compilerSettings.versions !== 'object') {
            throw Error(`Expected value of config.compilerSettings.versions to be a dictionary\nGot:${JSON.stringify(compilerSettings.versions)}`);
        }
        Object.entries(compilerSettings.versions).forEach(([k, v]) => {
            if (typeof v !== 'string' || !v.match(/^\d+.\d+.\d+$/)) {
                throw Error(`Expected value of config.compilerSettings.versions.${k} to be a valid version number\nGot:${v}`);
            }
        });
    }
    const appConfig = json;
    assertDir(appConfig.contractsDir, 'contractsDir');
    assertStr(appConfig.artifactsDir, 'artifactsDir');
    assertStr(appConfig.contractAbstractionDir, 'contractAbstractionDir');
    assertBool(appConfig.useDockerisedSolc, 'useDockerisedSolc');
    assertCompilerSettings(appConfig.compilerSettings);
}
