// parity/api/provider/postMessage.js
'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _createClass = function () {function defineProperties(target, props) {for (var i = 0; i < props.length; i++) {var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);}}return function (Constructor, protoProps, staticProps) {if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;};}();function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}var PostMessage = function () {



  function PostMessage(token, destination) {_classCallCheck(this, PostMessage);_initialiseProps.call(this);
    this._token = token;
    this._destination = destination;

    window.addEventListener('message', this.receiveMessage, false);
  }_createClass(PostMessage, [{ key: 'addMiddleware', value: function addMiddleware()

    {
    } }]);return PostMessage;}();var _initialiseProps = function _initialiseProps() {var _this = this;this.id = 0;this._callbacks = {};this.

  send = function (method, params, callback) {
    var id = ++_this.id;

    _this._callbacks[id] = callback;
    _this._destination.postMessage({
      id: id,
      from: _this._token,
      method: method,
      params: params,
      token: _this._token },
    '*');
  };this.

  receiveMessage = function (_ref) {var _ref$data = _ref.data,id = _ref$data.id,error = _ref$data.error,from = _ref$data.from,token = _ref$data.token,result = _ref$data.result,origin = _ref.origin,source = _ref.source;
    if (from !== 'shell' || token !== _this._token) {
      return;
    }

    _this._callbacks[id](error, result);
    _this._callbacks[id] = null;
  };};exports.default = PostMessage;module.exports = exports['default'];
